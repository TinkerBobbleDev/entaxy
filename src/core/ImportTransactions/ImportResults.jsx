import React from 'react'
import _ from 'lodash'
import PropTypes from 'prop-types'
import { withStyles } from '@material-ui/core/styles'
import Typography from '@material-ui/core/Typography'
import Button from '@material-ui/core/Button'
import Divider from '@material-ui/core/Divider'
import CheckCircleIcon from '@material-ui/icons/CheckCircle'
import ErrorIcon from '@material-ui/icons/Error'
import green from '@material-ui/core/colors/green'
import red from '@material-ui/core/colors/red'
import Table from '@material-ui/core/Table'
import TableBody from '@material-ui/core/TableBody'
import TableCell from '@material-ui/core/TableCell'
import TableHead from '@material-ui/core/TableHead'
import TableRow from '@material-ui/core/TableRow'

const styles = () => ({
  root: {
    padding: '20px'
  },
  formActions: {
    display: 'flex',
    'justify-content': 'flex-end',
    'padding-top': '10px'
  },
  iconCheck: {
    color: green[500],
    marginRight: '8px',
    'vertical-align': 'text-bottom'
  },
  iconError: {
    color: red[800],
    marginRight: '8px',
    'vertical-align': 'text-bottom'
  }
})

const ImportResults = ({
  classes,
  transactions,
  errors,
  onSave,
  onCancel
}) => {
  const maxErrorsToODisplay = 10
  const transactionTotalErrorsCount = Object.keys(errors.transactions).length
  const transactionErrorsToDisplay = Math.min(maxErrorsToODisplay, transactionTotalErrorsCount)
  const transactionErrorKeys = Object.keys(errors.transactions).slice(0, transactionErrorsToDisplay)

  return (
    <div>
      {errors.base.length === 0 && transactionErrorsToDisplay === 0 &&
        <div className={classes.root}>
          <Typography variant="h6" align="center">
            <CheckCircleIcon className={classes.iconCheck} />
            Found {transactions.length} transactions
          </Typography>
        </div>
      }
      {errors.base.length > 0 &&
        <div className={classes.root}>
          <Typography variant="h6" align="center">
            <ErrorIcon className={classes.iconError} />
            Some problems were found with the file
          </Typography>
          <Typography variant="body2" align="center">
            {errors.base[0]}
          </Typography>
        </div>
      }
      {transactionErrorsToDisplay > 0 &&
        <div className={classes.root}>
          <Typography variant="h6" align="center">
            <ErrorIcon className={classes.iconError} />
            Found errors in {transactionTotalErrorsCount} transactions.
          </Typography>
          {transactionTotalErrorsCount > transactionErrorsToDisplay &&
            <Typography variant="body" align="center">
              Displaying top {transactionErrorsToDisplay} errors
            </Typography>
          }
          <Table className={classes.table}>
            <TableHead>
              <TableRow>
                <TableCell>Line</TableCell>
                <TableCell>Error</TableCell>
              </TableRow>
            </TableHead>
            <TableBody>
              {_.map(transactionErrorKeys, line => (
                <TableRow key={line}>
                  <TableCell>{line}</TableCell>
                  <TableCell>{errors.transactions[line][0]}</TableCell>
                </TableRow>
              ))}
            </TableBody>
          </Table>
        </div>
      }
      <Divider />
      <div className={classes.formActions}>
        <Button
          onClick={onSave}
          color="secondary"
          disabled={errors.base.length > 0 || transactionErrorsToDisplay > 0}
        >
          Save Transactions
        </Button>
        <Button onClick={onCancel} color="secondary">Discard</Button>
      </div>
    </div>
  )
}

ImportResults.propTypes = {
  classes: PropTypes.object.isRequired,
  transactions: PropTypes.array.isRequired,
  errors: PropTypes.object.isRequired,
  onSave: PropTypes.func.isRequired,
  onCancel: PropTypes.func.isRequired
}

export default withStyles(styles)(ImportResults)
