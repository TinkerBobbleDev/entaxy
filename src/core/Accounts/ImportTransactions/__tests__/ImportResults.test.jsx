import React from 'react'
import renderer from 'react-test-renderer'
import ImportResults from '../ImportResults'

describe('ImportResults', () => {
  const mockOnSave = jest.fn()
  const mockOnCancel = jest.fn()

  beforeEach(() => {
    jest.clearAllMocks()
  })

  it('matches snapshot for BMO without errors', () => {
    const component = renderer.create((
      <ImportResults
        onSave={mockOnSave}
        onCancel={mockOnCancel}
        errors={{ base: [], transactions: [] }}
        transactions={[{ 0: {} }]}
        classes={{ }}
      />
    ))
    expect(component.toJSON()).toMatchSnapshot()
  })

  it('matches snapshot for BMO with errors on the header', () => {
    const component = renderer.create((
      <ImportResults
        onSave={mockOnSave}
        onCancel={mockOnCancel}
        errors={{ base: ['some error'], transactions: [] }}
        transactions={[{ 0: {} }]}
        classes={{ }}
      />
    ))
    expect(component.toJSON()).toMatchSnapshot()
  })

  it('matches snapshot for BMO with 1 error on a transaction', () => {
    const component = renderer.create((
      <ImportResults
        onSave={mockOnSave}
        onCancel={mockOnCancel}
        errors={{ base: [], transactions: [{ 0: ['some error'] }] }}
        transactions={[{ 0: {} }]}
        classes={{ }}
      />
    ))
    expect(component.toJSON()).toMatchSnapshot()
  })
})
