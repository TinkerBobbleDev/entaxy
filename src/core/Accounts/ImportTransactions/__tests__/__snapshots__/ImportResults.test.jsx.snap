// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImportResults matches snapshot for BMO with 1 error on a transaction 1`] = `
<div>
  <div
    className="ImportResults-root-1"
  >
    <h6
      className="MuiTypography-root-5 MuiTypography-h6-22 MuiTypography-alignCenter-28"
    >
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root-40 ImportResults-iconError-4"
        focusable="false"
        role="presentation"
        viewBox="0 0 24 24"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-2h2v2zm0-4h-2V7h2v6z"
        />
      </svg>
      Found errors in 
      1
       transactions.
    </h6>
    <table
      className="MuiTable-root-90"
    >
      <thead
        className="MuiTableHead-root-91"
      >
        <tr
          className="MuiTableRow-root-92 MuiTableRow-head-95"
        >
          <th
            aria-sort={null}
            className="MuiTableCell-root-97 MuiTableCell-head-98"
            scope="col"
          >
            Line
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root-97 MuiTableCell-head-98"
            scope="col"
          >
            Error
          </th>
        </tr>
      </thead>
      <tbody
        className="MuiTableBody-root-109"
      >
        <tr
          className="MuiTableRow-root-92"
        >
          <td
            aria-sort={null}
            className="MuiTableCell-root-97 MuiTableCell-body-99"
          >
            0
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root-97 MuiTableCell-body-99"
          >
            some error
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <hr
    className="MuiDivider-root-49"
  />
  <div
    className="ImportResults-formActions-2"
  >
    <button
      className="MuiButtonBase-root-80 MuiButtonBase-disabled-81 MuiButton-root-54 MuiButton-text-56 MuiButton-textSecondary-58 MuiButton-flat-59 MuiButton-flatSecondary-61 MuiButton-disabled-74"
      disabled={true}
      onBlur={[Function]}
      onClick={[MockFunction]}
      onContextMenu={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="-1"
      type="button"
    >
      <span
        className="MuiButton-label-55"
      >
        Save Transactions
      </span>
    </button>
    <button
      className="MuiButtonBase-root-80 MuiButton-root-54 MuiButton-text-56 MuiButton-textSecondary-58 MuiButton-flat-59 MuiButton-flatSecondary-61"
      disabled={false}
      onBlur={[Function]}
      onClick={[MockFunction]}
      onContextMenu={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="button"
    >
      <span
        className="MuiButton-label-55"
      >
        Discard
      </span>
      <span
        className="MuiTouchRipple-root-83"
      />
    </button>
  </div>
</div>
`;

exports[`ImportResults matches snapshot for BMO with errors on the header 1`] = `
<div>
  <div
    className="ImportResults-root-1"
  >
    <h6
      className="MuiTypography-root-5 MuiTypography-h6-22 MuiTypography-alignCenter-28"
    >
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root-40 ImportResults-iconError-4"
        focusable="false"
        role="presentation"
        viewBox="0 0 24 24"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-2h2v2zm0-4h-2V7h2v6z"
        />
      </svg>
      Some problems were found with the file
    </h6>
    <p
      className="MuiTypography-root-5 MuiTypography-body2-13 MuiTypography-alignCenter-28"
    >
      some error
    </p>
  </div>
  <hr
    className="MuiDivider-root-49"
  />
  <div
    className="ImportResults-formActions-2"
  >
    <button
      className="MuiButtonBase-root-80 MuiButtonBase-disabled-81 MuiButton-root-54 MuiButton-text-56 MuiButton-textSecondary-58 MuiButton-flat-59 MuiButton-flatSecondary-61 MuiButton-disabled-74"
      disabled={true}
      onBlur={[Function]}
      onClick={[MockFunction]}
      onContextMenu={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="-1"
      type="button"
    >
      <span
        className="MuiButton-label-55"
      >
        Save Transactions
      </span>
    </button>
    <button
      className="MuiButtonBase-root-80 MuiButton-root-54 MuiButton-text-56 MuiButton-textSecondary-58 MuiButton-flat-59 MuiButton-flatSecondary-61"
      disabled={false}
      onBlur={[Function]}
      onClick={[MockFunction]}
      onContextMenu={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="button"
    >
      <span
        className="MuiButton-label-55"
      >
        Discard
      </span>
      <span
        className="MuiTouchRipple-root-83"
      />
    </button>
  </div>
</div>
`;

exports[`ImportResults matches snapshot for BMO without errors 1`] = `
<div>
  <div
    className="ImportResults-root-1"
  >
    <h6
      className="MuiTypography-root-5 MuiTypography-h6-22 MuiTypography-alignCenter-28"
    >
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root-40 ImportResults-iconCheck-3"
        focusable="false"
        role="presentation"
        viewBox="0 0 24 24"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm-2 15l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
        />
      </svg>
      Found 
      1
       transactions
    </h6>
  </div>
  <hr
    className="MuiDivider-root-49"
  />
  <div
    className="ImportResults-formActions-2"
  >
    <button
      className="MuiButtonBase-root-80 MuiButton-root-54 MuiButton-text-56 MuiButton-textSecondary-58 MuiButton-flat-59 MuiButton-flatSecondary-61"
      disabled={false}
      onBlur={[Function]}
      onClick={[MockFunction]}
      onContextMenu={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="button"
    >
      <span
        className="MuiButton-label-55"
      >
        Save Transactions
      </span>
      <span
        className="MuiTouchRipple-root-83"
      />
    </button>
    <button
      className="MuiButtonBase-root-80 MuiButton-root-54 MuiButton-text-56 MuiButton-textSecondary-58 MuiButton-flat-59 MuiButton-flatSecondary-61"
      disabled={false}
      onBlur={[Function]}
      onClick={[MockFunction]}
      onContextMenu={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="button"
    >
      <span
        className="MuiButton-label-55"
      >
        Discard
      </span>
      <span
        className="MuiTouchRipple-root-83"
      />
    </button>
  </div>
</div>
`;
